B-21941 INT - I-13733-I-13739 TOO Queue Destination Requests Filtering Updates

traskowskycaci opened the pull request:
## [Agility ticket](https://www13.v1host.com/USTRANSCOM38/assetdetail.v1?number=B-21941)
First INT PR: https://github.com/transcom/mymove/pull/14465/files

Merged in as upstream but then reverted:
- 21940
original commit for merging in B-21940: https://github.com/transcom/mymove/pull/14464/commits/13a59c000a074864e515e1af0b532973f6328ffc
revert to remove the B-21940 changes from this branch: https://github.com/transcom/mymove/pull/14464/commits/2b4d5fe09c4126412a210b761401b6558b474149

Merged in upstream branches:
- 22272 
main branch: https://github.com/transcom/mymove/tree/MAIN-B-22272
merge commit: [Merge remote-tracking branch 'origin/MAIN-B-22272' into B-21941-MAIN-v2](https://github.com/transcom/mymove/pull/14464/commits/a890a0b1808812369b2392821ccb22c7855ee637)

- 22451
main branch: https://github.com/transcom/mymove/tree/MAIN-B-22451
merge commit: [Merge remote-tracking branch 'origin/MAIN-B-22451' into B-21941-MAIN-v2](https://github.com/transcom/mymove/pull/14464/commits/c87cfa2dba855fc97ab88c938fff34679584db06)

## Test moves:

**Please go through the process of making an actual move** if you use a test harness move and it's not appearing in the origin or destination queues as expected or acting weird. 

I found the test harness moves to not be 100% reliable, so I ended up making my own test moves for these - and will keep the moves locally and can huddle if any questions come up.

- One with only destination address update - should NOT appear in TOO queue and should appear in destination queue: 66K9TX
destination: 
![image](https://github.com/user-attachments/assets/9a1f9974-2515-49af-b3d2-9a5e71ccdf23)
origin:
![image](https://github.com/user-attachments/assets/bee15ad5-21f1-4097-9ddc-6276f023696e)

- One that has only origin SIT - should appear in TOO queue
destination: HD7746
![image](https://github.com/user-attachments/assets/cb4f0517-4ca0-4f12-9e98-e2e4bd662923)
origin: 
![image](https://github.com/user-attachments/assets/a1dd9e8d-e1d3-4c84-9d89-87036231cfe8)

- One that has only dest SIT - should NOT appear in TOO queue and should appear in destination queue: PJJQR7
destination:
![image](https://github.com/user-attachments/assets/b2623ece-3842-462a-af21-8ed0bb51ff49)

origin: 
![image](https://github.com/user-attachments/assets/2a70ff81-cd3a-4ed2-8090-3c573f25a0d2)

- One that has BOTH origin and dest SIT - should appear in TOO queue and should appear in destination queue: V6PGXF
destination: 
![image](https://github.com/user-attachments/assets/c6ed997f-0734-4086-b58f-76e6e0f5dcd8)

origin: 
![image](https://github.com/user-attachments/assets/13586156-f202-4975-8d99-2434a1c94b9c)

- One that has only domestic origin shuttle - should appear in TOO queue: DJKD8R
destination: 
![image](https://github.com/user-attachments/assets/d216766f-97ed-4be7-aeea-73f6081db63e)

origin: 
![image](https://github.com/user-attachments/assets/4e28a753-abd2-41e7-b257-314915fab211)

- One that has only dest shuttle - should NOT appear in TOO queue and should appear in destination queue: 48PX4B
![image](https://github.com/user-attachments/assets/f4ad8b53-549c-4f5f-b7b2-4d0a9ce62362)

origin: 
![image](https://github.com/user-attachments/assets/da0dc831-c66c-4b0f-b28b-26469584fd42)

- One that has BOTH origin and dest shuttle - should appear in TOO queue and should appear in destination queue: YM6KHF
destination: 
![image](https://github.com/user-attachments/assets/9594c777-ef40-4e9a-a52d-c57aed0cfc58)

origin: 
![image](https://github.com/user-attachments/assets/a0527b4f-ef2e-44ef-a4e3-62e8b16fb3f1)

- One that has only intl origin shuttle - should appear in TOO queue: 9BM93K
destination: 
![image](https://github.com/user-attachments/assets/5b7431fd-05c7-4317-b69a-a2bfa8992713)

origin: 
![image](https://github.com/user-attachments/assets/0a7508eb-fce6-4150-9156-0c114e136e38)

- One that has only intl destination shuttle - should appear in dest queue: MGR43H
destination: 
![image](https://github.com/user-attachments/assets/2f72f047-6a73-4c82-a3e8-7c61910293ba)

origin:
![image](https://github.com/user-attachments/assets/e08c09c3-8a61-456a-bcb1-0aa477193a8c)

- One that has BOTH intl origin and intl dest shuttle - should appear in TOO queue and should appear in destination queue: BMQHX4
destination: 
![image](https://github.com/user-attachments/assets/42ccaa4e-174c-4d11-a298-01c722165223)

origin: 
![image](https://github.com/user-attachments/assets/c769a305-64c6-4ca3-91ba-1405da6aa16a)

- One that has a SIT extension and only origin SIT - should appear in TOO queue: D3PBPJ
destination: 
![image](https://github.com/user-attachments/assets/eb7f0434-ced2-45ee-bf06-6eeb74044a03)

origin: 
![image](https://github.com/user-attachments/assets/691b38fa-2c8e-46fe-98dd-49161a1fbbc4)

- One that has a SIT extension and only dest SIT - should appear in dest queue: KCMJVB
destination: 
![image](https://github.com/user-attachments/assets/a6da9d4d-8799-497e-919d-6b59587b9fae)

origin: 
![image](https://github.com/user-attachments/assets/69d9f81d-468c-41d4-93c7-106e05027e4e)

- One that has a SIT extension and both origin and dest SIT - should appear in both queues: QGFJ7X
destination:
![image](https://github.com/user-attachments/assets/d2773077-b588-453d-bd26-32a7f296841b)

origin:
![image](https://github.com/user-attachments/assets/12d48616-b97c-4164-9928-bc437b82c9c7)

- One that has excess weight - should appear in TOO queue: KWH4RD
destination: 
![image](https://github.com/user-attachments/assets/2001a800-ac8d-4dcb-a419-475c5c318137)

origin: 
![image](https://github.com/user-attachments/assets/fb334de2-a1ed-4bd7-a131-2eb8a5711add)

- One that has excess UB weight - should appear in TOO queue: QT844H
destination: 
![image](https://github.com/user-attachments/assets/5df145ae-4262-4544-aa8f-f463417aaeb6)
origin: 
![image](https://github.com/user-attachments/assets/5fb7f9e8-945f-49bf-b600-076d82b59eaa)

- Test that the moves with excess weight only show up in the expected GBLOC office user TOO queue and not every other GBLOC's TOO queue: (relic of the old query that was an issue but should not be now)
shown in KKFA queue above
not in BGNC:
![image](https://github.com/user-attachments/assets/dbc89247-baec-4df7-a3f5-61782c8a496d)
![image](https://github.com/user-attachments/assets/4a292869-c13e-4835-b4e6-005d634cdf4f)

not in CCNQ:
![image](https://github.com/user-attachments/assets/cfddba71-179d-4979-8b34-76e58a82e0cf)
![image](https://github.com/user-attachments/assets/a3a82907-29d0-4d33-80e1-933472c5ac11)

## Changes since last review: logic around SIT extensions
- One that has a SIT extension and only origin SIT - should appear in TOO queue: D3PBPJ
destination: 
![image](https://github.com/user-attachments/assets/eb7f0434-ced2-45ee-bf06-6eeb74044a03)

origin: 
![image](https://github.com/user-attachments/assets/691b38fa-2c8e-46fe-98dd-49161a1fbbc4)

- One that has a SIT extension and only dest SIT - should appear in dest queue: KCMJVB
destination: 
![image](https://github.com/user-attachments/assets/a6da9d4d-8799-497e-919d-6b59587b9fae)

origin: 
![image](https://github.com/user-attachments/assets/69d9f81d-468c-41d4-93c7-106e05027e4e)

- One that has a SIT extension and both origin and dest SIT - should appear in both queues: QGFJ7X
destination:
![image](https://github.com/user-attachments/assets/d2773077-b588-453d-bd26-32a7f296841b)

origin:
![image](https://github.com/user-attachments/assets/12d48616-b97c-4164-9928-bc437b82c9c7)

traskowskycaci commented:
Just one flaky test failure, still good to review

TevinAdams reviewed (dismissed):
Created a few different moves and didn't find any issues with them.

traskowskycaci commented:
Added a commit to cleanup some test comments, rerequesting @TevinAdams's review. What I added:

- cleanup commit, no change to functionality: https://github.com/transcom/mymove/pull/14586/commits/cdad4c6f70c85da33ca02caa5f9c1ceac0989f25
- merge of main to int: https://github.com/transcom/mymove/pull/14586/commits/588caaf669188b7e1b4765a4b4bd9534fe1e55b8

danieljordan-caci commented:
Per our convo/huddle in Slack, this will be approved once the query is updated to:
```
			NOT (
				(
					-- find moves with destination requests (submitted destination re_services codes)
					EXISTS (
						SELECT 1
						FROM mto_service_items msi
						JOIN re_services rs ON msi.re_service_id = rs.id
						WHERE msi.mto_shipment_id = mto_shipments.id
						AND msi.status = 'SUBMITTED'
						AND rs.code IN ('DDFSIT', 'DDASIT', 'DDDSIT', 'DDSHUT', 'DDSFSC', 'IDFSIT', 'IDASIT', 'IDDSIT', 'IDSHUT')
					)
					-- moves with destination requests (destination address update requested)
					OR
					EXISTS (
						SELECT 1
						FROM shipment_address_updates sau
						WHERE sau.shipment_id = mto_shipments.id
						AND sau.status = 'REQUESTED'
					)
				)
				-- find moves with no origin requests
				-- find moves without submitted destination re_services codes
				AND NOT EXISTS (
					SELECT 1
					FROM mto_service_items msi
					JOIN re_services rs ON msi.re_service_id = rs.id
					WHERE msi.mto_shipment_id = mto_shipments.id
					AND msi.status = 'SUBMITTED'
					AND rs.code NOT IN ('DDFSIT', 'DDASIT', 'DDDSIT', 'DDSHUT', 'DDSFSC', 'IDFSIT', 'IDASIT', 'IDDSIT', 'IDSHUT')
				)
				-- find moves without excess weight risk to acknowledge
				AND NOT (
					moves.excess_weight_qualified_at IS NOT NULL 
					AND moves.excess_weight_acknowledged_at IS NULL
				)
				-- find moves without UB excess weight risk to acknowledge
				AND NOT (
					moves.excess_unaccompanied_baggage_weight_qualified_at IS NOT NULL 
					AND moves.excess_unaccompanied_baggage_weight_acknowledged_at IS NULL
				)
			)
```

was the age old issue of excess weight, but seems to be fixed with this query. Nice job! I'll await the next commit with paired tests.

danieljordan-caci reviewed (commented):
![Screenshot 2025-01-16 at 11 29 35 AM](https://github.com/user-attachments/assets/52e3b7c0-6652-4092-8dcf-aba4bfa3d100)

Tested several moves:
- only origin SIT (showed up in TOO queue)
- only dest SIT (did not show in TOO queue, but showed in dest)
- both o & d SIT (showed up in both)
- only origin shuttle (showed up in TOO queue)
- only dest shuttle (showed up in dest queue only)
- both shuttles (both queues)
- excess weight with only destination (showed up in TOO queue)
- excess UB weight with destination SIT (showed up in TOO queue - although couldn't approve SIT)

Exhausting work, I'm sure. Awaiting next commit, will verify query has changed to what I have that's working.

traskowskycaci commented:
> Per our convo/huddle in Slack, this will be approved once the query is updated to:
> 
> ```
> 			NOT (
> 				(
> 					-- find moves with destination requests (submitted destination re_services codes)
> 					EXISTS (
> 						SELECT 1
> 						FROM mto_service_items msi
> 						JOIN re_services rs ON msi.re_service_id = rs.id
> 						WHERE msi.mto_shipment_id = mto_shipments.id
> 						AND msi.status = 'SUBMITTED'
> 						AND rs.code IN ('DDFSIT', 'DDASIT', 'DDDSIT', 'DDSHUT', 'DDSFSC', 'IDFSIT', 'IDASIT', 'IDDSIT', 'IDSHUT')
> 					)
> 					-- moves with destination requests (destination address update requested)
> 					OR
> 					EXISTS (
> 						SELECT 1
> 						FROM shipment_address_updates sau
> 						WHERE sau.shipment_id = mto_shipments.id
> 						AND sau.status = 'REQUESTED'
> 					)
> 				)
> 				-- find moves with no origin requests
> 				-- find moves without submitted destination re_services codes
> 				AND NOT EXISTS (
> 					SELECT 1
> 					FROM mto_service_items msi
> 					JOIN re_services rs ON msi.re_service_id = rs.id
> 					WHERE msi.mto_shipment_id = mto_shipments.id
> 					AND msi.status = 'SUBMITTED'
> 					AND rs.code NOT IN ('DDFSIT', 'DDASIT', 'DDDSIT', 'DDSHUT', 'DDSFSC', 'IDFSIT', 'IDASIT', 'IDDSIT', 'IDSHUT')
> 				)
> 				-- find moves without excess weight risk to acknowledge
> 				AND NOT (
> 					moves.excess_weight_qualified_at IS NOT NULL 
> 					AND moves.excess_weight_acknowledged_at IS NULL
> 				)
> 				-- find moves without UB excess weight risk to acknowledge
> 				AND NOT (
> 					moves.excess_unaccompanied_baggage_weight_qualified_at IS NOT NULL 
> 					AND moves.excess_unaccompanied_baggage_weight_acknowledged_at IS NULL
> 				)
> 			)
> ```
> 
> was the age old issue of excess weight, but seems to be fixed with this query. Nice job! I'll await the next commit with paired tests.

Found that this query is over-filtering (filtering out a move with origin SIT and excess weight) so still working on the final solution. Marking PR as draft for now.

cameroncaci reviewed (commented):
Working good, what about uncrating though? I see that if uncrating is present it doesn't go into destination queue. Would uncrating be considered dest?

danieljordan-caci commented:
> Working good, what about uncrating though? I see that if uncrating is present it doesn't go into destination queue. Would uncrating be considered dest?

Nope. This is from the feature:
```
destination GBLOC will be responsible for approving destination SIT and shuttle service items.
```

cameroncaci reviewed (dismissed):
Worked for me and meets AC - love the new query

traskowskycaci commented:
Closing due to polluted git history. Cherry-picked commits into new main and int branch. New integration PR here: https://github.com/transcom/mymove/pull/14810
